doc/proposals/shared-memory-transport/eprosima/interprocess_shared_mem.md

```diff
@@ -0,0 +1,95 @@
+
+* **SharedMemoryManager**: Applications instantiate this object in order to access the shared memory resources described above. At least one per process memory manager is required. The manager provides functionality for the application to create shared memory segments, alloc data buffers in the segments, push buffer descriptors to shared memory ports and create listeners associated to a port.
+
+### Example scenario
+![](interprocess_shared_mem1.png)
+
+Let's study the above example. There are three processes with a SharedMemManager per process, every process creates its own shared memory segment intended to store the buffers that will be shared with other processes.
+
+P1, P2 and P3 processes are participants in a RTPS-DDS environment. Discovery of participants is done using multicast so we have selected shared memory port0 as "multicast" port for discovery, therefore, first thing all participants do is to open shared memory port0. A listener attached to port0 is created too, by every participant, to read the incoming descriptors.
+
+Each participant opens a port to receive unicast messages, ports 1, 2 and 3 respectively, and create listeners associated to those ports.
+
+The first message the participants send is the multicast discovery message: "I'm here, and I am listening on portN". So they alloc a buffer in its local segment, write the message to that buffer and push the buffer's descriptor through the port0. Observe, in the global segment, how port0's ring-buffer store the descriptors to Data1a(P1), Data2a(P2), Data3a(P3). 
+
+After the discovery phase, participants know the other participants and its "unicast" ports so they can push messages to specific participants.
+
+Finally, let's observe how P1 share Data1b with P3 and Data1c with P2 and P3, this is done by pushing the buffer's descriptors to the corresponding ports, the descriptors are stored in the port's ring-buffer and listeners registered in these ports are notified.
+
+### Design considerations
+
+* **Minimize global interprocess locks**: Interprocess locks are dangerous because, in the case of one of the involved processes crash or hung while holding an interprocess lock, all the collaborating processes could be affected. In this design, the global segment is the critical area, most of accesses will be lock-free reading operations. Interprocess-locks will only be required when registering / unregistering a new listener in a port. Once the setup has been established, the risk of a process locking the whole port is minimized.
```

Comment:
In a shared-memory port's segment resides a RingBuffer, an interprocess-mutex, and an interprocess-condition variable. Operations over the ring buffer are lock-free atomic (compare and swap) so writers and readers may push or pop data without locking any mutex. 


The mutex and condition variable are used to wait on when the ring buffer is empty. So if a reader detects the ring buffer is empty it will wait on the condition. When writers push elements, lock the empy condition before pushing and after pushing notify on the condition to wake-up listeners.

So if no bugs or crashes everything is OK, but if a process crash holding an interprocess-lock it will block all readers and writers using that port. This is bug situation, so involved processes would need to be restarted or a fail-tolerance mecanishm is needed.